### ISSUE TITLE
Shake report is reporting negative build times for many rules 

### ISSUE THREAD

**Issue by mavnn** [ndmitchell/shake#752]
I've recently turned on Shake profiling for a project, and am a bit surprised to see that it would take a negative amount of time to compile single threaded:

![image](https://user-images.githubusercontent.com/309312/79970905-88489480-848b-11ea-8b03-0b36ae671e13.png)

Alas, experimental verification concludes that you haven't invented a time machine.

Interestingly, no commands are shown to take a negative time, but plenty of rules are:

![image](https://user-images.githubusercontent.com/309312/79971051-ba59f680-848b-11ea-9664-3c4207afd514.png)

I can't share the project in question, but I was wondering if you had any thoughts on how I could investigate and if there is further debugging information I could provide.

**Comment by ndmitchell** [comment]
Does your project use `parallel` at all? That used to duplicate counts in some cases and could lead to negative numbers.  I don't remember when I fixed it, but more recent versions have better answers. Which version are you on?

**Comment by mavnn** [comment]
We do use `parallel`, quite a lot in fact (including in those rules). Version is 0.18.4.

I'll investigate that as a possible cause!

**Comment by ndmitchell** [comment]
Looking through the changelog allegedly the fixes for parallel time recording landed in 0.16, so these may be fresh time recording bugs. The place to look is probably near https://github.com/ndmitchell/shake/blob/master/src/Development/Shake/Internal/Core/Action.hs#L499

**Comment by mavnn** [comment]
Removing `parallel` from our rules hasn't fixed the issue, but it has highlighted that this only appears to crop up in rules that use `batch` so I'm going to have a look there and see what I can find.

**Comment by ndmitchell** [comment]
See 8c245cb for the fix and 29e0642 for the test. The code before was pretty wrong on a few dimensions. The main problem was we never timed the action we were running, so instead of taking the time spend and dividing by the number of entries in the batch, we took the amount to discount by and divided that by the number of entries. We also discounted by the amount of time spent running. The maths for 2 jobs, where the rules together take 10 seconds, of which 6 is waiting for a dependency, would be a discount of:

> 10 + (6 / 2) = 13

And given it probably takes 10s to run in total (assuming the pre-batch action is cheapish), then you get a discount of 13s and a result of -3s. Oops.

Now we actually measure the inner job, subtract the inner discount, divide by the number, and _add_ that cost (as a negative discount). Much better!

Since I've screwed up both parallel and batch many times, I added a pretty good test suite, so hopefully that's the last time!
